plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/InstanceOfCheckCastBasicBlockBuilder.java

```diff
@@ -189,15 +189,22 @@ public Value instanceOf(final Value input, final ObjectType expectedType, int ex
 
     public Value classOf(final Value typeId) {
         MethodElement methodElement = ctxt.getVMHelperMethod("classof_from_typeid");
+
         if (typeId instanceof TypeLiteral) {
             ValueType valueType = ((TypeLiteral) typeId).getValue();
             if (valueType instanceof ReferenceType) {
                 ctxt.error(getLocation(), "Class of reference type");
                 return ctxt.getLiteralFactory().zeroInitializerLiteralOfType(methodElement.getType().getReturnType());
+            } else if (valueType instanceof ReferenceArrayObjectType) {
+                ReferenceArrayObjectType referenceArrayType = (ReferenceArrayObjectType) valueType;
+                Value elementTypeId = ctxt.getLiteralFactory().literalOfType(referenceArrayType.getLeafElementType());
+                int dims = referenceArrayType.getDimensionCount();
+                List<Value> args = List.of(elementTypeId, ctxt.getLiteralFactory().literalOf(ctxt.getTypeSystem().getUnsignedInteger8Type(), dims));
+                return notNull(getFirstBuilder().call(getFirstBuilder().staticMethod(methodElement), args));
```

Comment:
A type literal can never carry a type of `ReferenceArrayObjectType`.  This is because the type ID of _any_ reference array class is always going to be the same: the type ID of the special internal reference array implementation class.  That means that the `typeIdOf` node won't actually work correctly for that case.

The logic for actually finding the array type should probably reside completely within the helper method.